server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.cluster.node.shutdown;
+
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.support.master.MasterNodeReadRequest;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.List;
+
+public class PrevalidateNodeRemovalRequest extends MasterNodeReadRequest<PrevalidateNodeRemovalRequest> {
+
+    private final String[] nodeIds; // Maybe named just nodes?
+    private DiscoveryNode[] concreteNodes;
+
+    public PrevalidateNodeRemovalRequest(String... nodeIds) {
+        this.nodeIds = nodeIds;
+    }
+
+    public PrevalidateNodeRemovalRequest(final StreamInput in) throws IOException {
+        super(in);
+        nodeIds = in.readStringArray();
+    }
+
+    public void setConcreteNodes(DiscoveryNode... discoveryNodes) {
+        concreteNodes = discoveryNodes;
```

Comment:
I wonder why we set this into the request, can we not keep that in a local state instead? I prefer the request to be purely the request and also that writeTo and reading results in the same object.